desc/coils.py

```diff
@@ -1123,43 +1162,54 @@ def from_makegrid_coilfile(cls, coil_file, method="cubic"):
                     and headind != -1
                 ):
                     coilinds.append(i)
-                    names.append(" ".join(line.split()[4:]))
-        if len(lines[3 + headind].split()) != 4:
-            raise OSError(
-                "4th line in file must be the start of the first coil! "
-                + "Expected a line of length 4 (after .split()), "
-                + f"instead got length {lines[3].split()}"
-            )
-        header_lines_not_as_expected = np.array(
-            [
-                len(lines[0 + headind].split()) != 2,
-                len(lines[1 + headind].split()) != 2,
-                len(lines[2 + headind].split()) != 2,
-            ]
-        )
-        if np.any(header_lines_not_as_expected):
-            raise OSError(
-                "First 3 lines in file starting with the periods line "
-                + "must be the header lines,"
-                + " each of length 2 (after .split())! "
-                + f"Line(s) {lines[np.where(header_lines_not_as_expected)[0]+headind]}"
-                + " are not length 2"
-            )
-
-        for i, (start, end) in enumerate(zip(coilinds[0:-1], coilinds[1:])):
+                    groupname = " ".join(line.split()[4:])
+                    groupind = int(groupname.split()[0].strip())
+                    if groupind not in coils.keys():
+                        coils[groupind] = []
+                        groupnames.append(groupname)
+                    coilnames.append(groupname)
+                    groupinds.append(groupind)
+
+        for i, (start, end, groupind, coilname) in enumerate(
+            zip(coilinds[0:-1], coilinds[1:], groupinds, coilnames)
+        ):
             coords = np.genfromtxt(lines[start + 1 : end])
-            coils.append(
+            coils[groupind].append(
                 SplineXYZCoil(
                     coords[:, -1][0],
                     coords[:, 0],
                     coords[:, 1],
                     coords[:, 2],
                     method=method,
-                    name=names[i],
+                    name=coilname,
                 )
             )
-
-        return cls(*coils)
+        # if it is a single group, then only return one coilset, not a
+        # nested CoilSet
+        groupinds = list(coils.keys())
+        class_init = cls
+        try:
+            if len(groupinds) == 1:
+                return class_init(*coils[groupinds[0]], name=groupnames[0])
+            else:  # return a nested coilset, containing one coilset per coilgroup
+                return MixedCoilSet(
```

Comment:
I'm not sure this is the best way. Since it's a class method, it should always return an instance of the class that called it, while this may sometimes return a `MixedCoilSet` even if called like `CoilSet.from_mgrid_file`

I think what we want is something more like 
- if called from `CoilSet`, try to return a single coilset with all the coils, ignoring groups. If we can't due to different number of knots etc, throw an error and tell user to try using MixedCoilSet
- If called from `MixedCoilSet`, break into groups and return a `MixedCoilSet` of where each subgroup is a `CoilSet` if possible, falling back to `MixedCoilSet` for the subgroups if necessary.
- May also want a an option to ignore groups when loading `MixedCoilSet`? Not as sure about that though.

Reply:
I think that `CoilSet` works if each of the individual CoilSets are `CoilSet` classes though.

A flatten option would be easy, to just return as a single coilset objet

Reply:
I can do this, but I don't love that the user might receive an error when they go to load a CoilSet, we could make this a separate method instead of a class method?

Reply:
As it is right now there are coilsets that can't be loaded with `CoilSet.from_mgrid_file` so I don't think it's really an issue. As long as the error says why it happened and how to fix it (using `MixedCoilSet`) i think its fine